### DISCUSSION TITLE
Problems with new Specialist-Army command

### DISCUSSION THREAD

**Discussion by Anprivate** [fedorovvl/tso_client#195] [General]
At first - tyvm, it is long waited command.
Some problems.
1. List of generals is not sorted. For example - I have 15 ghosts, all renamed as Ghost 01 - 15 and order of them in list is very strange https://prnt.sc/06AUd2KuyFDq
2. Generals loaded, template saved. Ok, load template, load generals, returning to the main list - loading of some newly loaded generals is not changed. Not all generals, only some. https://prnt.sc/0Xmn0FoqNpWe https://prnt.sc/beECbnhjRSHO
3. Sorting of troops in generals is strange - not recruits-bowmens-militia... but cavalry - recruits - elite soldiers 
4. After unloading generals disappeared from list.

**Comment by fedorovvl** [comment]
1. generals was sorted by its original name
2. between loading army (server req) and changes in client there is some time lag.. table build earlyer that client receive changes from server. don't know how to fix yet
3. Is it important?
4. after last general unloaded list refreshes.. or not?

**Reply by Anprivate** [reply]
1. It is very uncomfortable when you have 15 ghosts / 9 Younger etc :)
2. Ok, no problem.
3. Just unusual - hard to fast check loading with eyes. No problem
4. Strange, Now I checked - yes, refreshed after last unload. But I remember - it was list with skipped generals. Strange. Will check more.

**Reply by fedorovvl** [reply]
1. check on latest build
2. check on latest build
3. check on latest build

**Comment by Anprivate** [comment]
Just checked a new version - tyvm (а нельзя добавить возможность доната через СБП? А то вбивать полные данные карточки обычно лень, а через СБП и QR код это делается в пару кликов :) ).
1. Great, now generals are sorted. But I can't understand order of sorting. I suppose, it have to be traditional ascending alphabetical sorting, but in fact - it something else - quartermaster -> champion -> medic ->General Boris. How is it possible - I can't understand. https://prnt.sc/EO2QsESyQp2z
2. Still same problem - see YGG 3 - in fact is loaded with 155 recruits, but https://prnt.sc/gPTtIqyoROXs
3. Nothing changed https://prnt.sc/3QuxKuOHOl0A
4. I caught it - generals disappeared from list after unloading and no refresh https://prnt.sc/cR4ftOhHnaJz

And may be recommendation about this part - https://prnt.sc/SFIPa2Z_-efQ
Make two lines - for normal and elite troops and do sorting of generals in same order as in normal general troops window:
normal: recruits - bowman - militia - cavalry - longbowman - soldier - crossbowman - elite soldier - cannoneer
elite: swordsman - mounted swordsman - knight - marksmen - armoured marksmen - mounted marksmen - besieger
It will be much easier to read it

**Reply by fedorovvl** [reply]
u test on 1.5.5.1?

**Reply by Anprivate** [reply]
Sure https://prnt.sc/ZNFGXoo80YlA

**Reply by fedorovvl** [reply]
unloading process looks like progressbar -> watches for 3 sec -> reload ?

**Reply by Anprivate** [reply]
Usually - yes. And in this situation generals are not disappeared. But sometimes - after 3 sec - no reload and generals still not in list.
It is not often situation, but sometimes it happens.

**Reply by fedorovvl** [reply]
check system chat after that, if error occurs message will be posted in system chat

**Reply by Anprivate** [reply]
And another problem sometimes - after loading generals window sometimes not returning to full generals list - https://prnt.sc/8G77PBSd-8ze
And - sometimes after unloading and normal refreshing unloaded general still indicated as loaded (ghost 4) - https://prnt.sc/sIICf_xmWy00

**Reply by Anprivate** [reply]
Error loading TypeError: Result of expression 'item.split' [undefined] is not a function.

**Reply by Anprivate** [reply]
But I can't reproduce error, so I don't know - in what exact moment this error was. Will try more.

**Reply by Anprivate** [reply]
New
[13:02] Client [army]: Error loading TypeError: Result of expression 'spec.GetGarrison' [undefined] is not a function.
When loading template

**Reply by fedorovvl** [reply]
try on latest test build
new sorting
separate free army units
some error handling

**Reply by Anprivate** [reply]
Where can I get it?

**Reply by fedorovvl** [reply]
"latest build" badge on main repo page
autobuilded on push

**Reply by Anprivate** [reply]
Done
After unload last general sometimes still indicated as loaded - ghost 4 is unloaded in fact - https://prnt.sc/IwfGaOnqQfd-
And same after loading with template - sometimes last general loading indication is incorrect.
Troops order in generals is great now. But not in loading window - https://prnt.sc/OBdNcuh7c_Jl

Will try more and write here any new information.

**Reply by Anprivate** [reply]
And again
 [13:49] Client [army]: Error loading TypeError: Result of expression 'spec.GetGarrison' [undefined] is not a function.
When I'm trying to open file with template

**Reply by fedorovvl** [reply]
i increased wait time to 5 sec.. maybe its help (depends on internet speed maybe)
latest build

**Reply by Anprivate** [reply]
Ok, I've tried it and 5 sec is enough for correct refresh now.
And yes, internet speed is not good here (turkish hotel's wifi).
About sorting in loading window - no problem. I just resaved templates with sorted generals and now in loading window is the same order.

**Comment by fedorovvl** [comment]
pls try again on latest build
now it waits all packets before refresh

**Reply by Anprivate** [reply]
Done
Still got error sometimes when loading template - [14:39] Client [army]: Error loading TypeError: Result of expression 'spec.getName' [undefined] is not a function.
Loading looks correct

**Comment by Anprivate** [comment]
Just idea - to check current load of generals before reloading and doesn't reload troops for generals, which are already loaded correctly.
For example - I have already loaded team of generals, ready to adventure. I do "load template" and in current version all this generals will be reloaded even if current loading is correct. It is not necessary and just wasting of time.

**Reply by fedorovvl** [reply]
implemented in https://github.com/fedorovvl/tso_client/commit/72e6b796fbf446cc6d760c4825a95b3c356f3703
also required army calculation changed
try last build

**Reply by Anprivate** [reply]
Great, tyvm. Can you change "Close" button on loading template page to "return" or "back" - to return to main Army page? If all generals are already loaded correctly I just can't send they to adventure - I have to press "close" and then open Army menu again.

**Reply by fedorovvl** [reply]
click reset If all generals are already loaded correctly

**Reply by Anprivate** [reply]
But after reset all generals will be unselected - I will have to reselect generals which I want to send

**Reply by fedorovvl** [reply]
not anymore https://github.com/fedorovvl/tso_client/commit/d0a676c4e8b554e6eb968892ca76bfb0c478b944

**Reply by Anprivate** [reply]
Hm. Still unselected after reset

**Reply by fedorovvl** [reply]
download latest build and check =)

**Reply by Anprivate** [reply]
Of course I did it. Think it was too fast :) Yes, now it is working. Tyvm (but "Reset" is slightly weird now - what reset this button?)
